
 #                                      eBay Kleinanzeigen
 

The aim of this project is to clean the data and analyze the included used car listings from a classifieds section of the German eBay website. The dataset can be found on data.world

In [245]:
import pandas as pd
import numpy as np

# Importing Pandas and Numpy libraries

In [246]:
# Reading the autos.csv file into pandas

autos = pd.read_csv('C:\\Users\koraim\\OneDrive - Dell Technologies\\Desktop\\Data Science\\Datasets\\autos.csv', encoding= "Windows-1252")

In [247]:
# Exploring the dataset, especially the first 5 rows

autos.head()
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [248]:
autos # Exploring the first and last five rows of pandas object "autos"

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

The dataset consists of 20 columns and 50,000 data points. Some columns contain null values like the vehicleType, powerPS, notRepairedDamage. Also the dataset naming convention is camel case instead of snake case.

In [249]:


autos.columns  # printing an array which consists of the exisiting column names

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']    # Columns names changed to snakecase


autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model  kilometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We started data cleaning by converting the column names from camelcase into snakecase.

We will now do some basic data explortion using the df.describe() and series.value_counts()  , we will look out for:

* Numeric data stored as text
* Columns where most/all the values are the same and see if we can drop them from our analysis.


In [250]:
autos.describe(include='all')
autos['price'].value_counts()

autos.rename(columns={'kilometer':'odemeter_km'}, inplace=True) # We changed the name of column kilometer to odemeter_km

autos["odemeter_km"].head()


0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odemeter_km, dtype: object

As we continue to observe our dataset here, it seems *seller* and *price* columns have only 2 unique values. Also the *nr_of_pictures* column has 0 values.

The *price* column has 1421 values of $0. We may want to leave out those rows before we prceed with our analysis.

So let's start by cleaning the **price** column:

In [251]:
#cleaning the price column:

autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()







0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

The **price** column is in good shape now, we shift towards the **odemeter** column and clean it as well:

In [252]:
#Cleanng the odemeter_km column:
autos["odemeter_km"] = (autos["odemeter_km"]
                          .str.replace("km","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["odemeter_km"].head()


0    150000
1    150000
2     70000
3     70000
4    150000
Name: odemeter_km, dtype: int32

We will now proceed by dropping columns(**seller, offer_type and nr_of_pictures**)`as they have same unique values or 0.

In [253]:
#Removing the seller, offer_type and nr_of_pictures columns from our dataset:

autos = autos.drop(["seller","offer_type","nr_of_pictures"],axis=1)





Let's continue exploring the dataset, we will now investigate the values itself and check if we have any unrealistic ones:

In [254]:

autos.describe()


price  registration_year      powerPS   odemeter_km  \
count     50,000.00000       50,000.00000 50,000.00000  50,000.00000   
mean       9,840.04376        2,005.07328    116.35592 125,732.70000   
std      481,104.38050          105.71281    209.21663  40,042.21171   
min            0.00000        1,000.00000      0.00000   5,000.00000   
25%        1,100.00000        1,999.00000     70.00000 125,000.00000   
50%        2,950.00000        2,003.00000    105.00000 150,000.00000   
75%        7,200.00000        2,008.00000    150.00000 150,000.00000   
max   99,999,999.00000        9,999.00000 17,700.00000 150,000.00000   

       registration_month  postal_code  
count        50,000.00000 50,000.00000  
mean              5.72336 50,813.62730  
std               3.71198 25,779.74796  
min               0.00000  1,067.00000  
25%               3.00000 30,451.00000  
50%               6.00000 49,577.00000  
75%               9.00000 71,540.00000  
max              12.00000 99,998.00000

As we continue to investigate the data, we start with the **price** column:

In [255]:
# checking the values of the price column:

print(autos['price'].unique().shape) #to see the unique values

autos['price'].describe() # to check max/min/mean

(2357,)


count       50,000.00000
mean         9,840.04376
std        481,104.38050
min              0.00000
25%          1,100.00000
50%          2,950.00000
75%          7,200.00000
max     99,999,999.00000
Name: price, dtype: float64

In [256]:
# changing the format to show floats with 1 decimal place:
pd.options.display.float_format = '{:,.1f}'.format
autos['price'].describe()

count       50,000.0
mean         9,840.0
std        481,104.4
min              0.0
25%          1,100.0
50%          2,950.0
75%          7,200.0
max     99,999,999.0
Name: price, dtype: float64

In [257]:
values = autos['price'].value_counts()
values.sort_index(ascending=False).head(20) # Sorting the values of column price in descending order

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

We have some values above 350,000 which could be considered as outliers. Also the the values above 1 million seem very odd.

In [258]:
values.sort_index(ascending=True).head(20) # Exploring the other way

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

We will consider any value less than 1$ and above 350,000 as outlier so proceeding with removing them:

In [259]:
autos = autos[autos["price"].between(1,350000)] #Keeping the price values which is between 1 and 350,000


Then, we will switch to the odemeter_km column and check its values and remove any outliers as well:

In [260]:
print(autos["odemeter_km"].value_counts().sort_index(ascending=True).head(30)) #Exploring the data in the odemeter column

pd.options.display.float_format = '{:,.1f}'.format

autos["odemeter_km"].describe()



5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odemeter_km, dtype: int64


count    48,565.0
mean    125,770.1
std      39,788.6
min       5,000.0
25%     125,000.0
50%     150,000.0
75%     150,000.0
max     150,000.0
Name: odemeter_km, dtype: float64

We will leave the **odemeter_km** as it is as I cant see any outlier value. Moving on now to the date columns and check their data ranges:

In [261]:
autos["registration_year"].describe()

count   48,565.0
mean     2,004.8
std         88.6
min      1,000.0
25%      1,999.0
50%      2,004.0
75%      2,008.0
max      9,999.0
Name: registration_year, dtype: float64

It seems we have unrelaistic year values here as the minimum is 1000. Maximum year value is 9999 which way ahead in the future as well so we will remove the outliers of the year range 1900-2016:

In [262]:
autos = autos[autos["registration_year"].between(1900,2016)] #keeping only years between 1900 and 2016

print(autos['registration_year'].describe())

count   46,681.0
mean     2,002.9
std          7.2
min      1,910.0
25%      1,999.0
50%      2,003.0
75%      2,008.0
max      2,016.0
Name: registration_year, dtype: float64


Now, we will have a look on which of the car brands have highest price, we will begin to explore the brand column as follows:

In [263]:
brands = autos["brand"].value_counts(normalize=True) # We check the brands of the listings, and how big is its representation
print(brands)
print(brands.index)

volkswagen       0.2
bmw              0.1
opel             0.1
mercedes_benz    0.1
audi             0.1
ford             0.1
renault          0.0
peugeot          0.0
fiat             0.0
seat             0.0
skoda            0.0
nissan           0.0
mazda            0.0
smart            0.0
citroen          0.0
toyota           0.0
hyundai          0.0
sonstige_autos   0.0
volvo            0.0
mini             0.0
mitsubishi       0.0
honda            0.0
kia              0.0
alfa_romeo       0.0
porsche          0.0
suzuki           0.0
chevrolet        0.0
chrysler         0.0
dacia            0.0
daihatsu         0.0
jeep             0.0
subaru           0.0
land_rover       0.0
saab             0.0
jaguar           0.0
daewoo           0.0
trabant          0.0
rover            0.0
lancia           0.0
lada             0.0
Name: brand, dtype: float64
Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan',

The German manufactures represent 50% of the top 10 brands in the car listings. VW is the most popular, followed by BMW, Opel, Mercedes and last but not least Audi. 

Lots of brands do not have significant percentage of listings, so we will remove any brand which have lessa than 5% representation in the following step.

In [264]:
common_brands = brands[brands > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [266]:
brands_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price'].mean()
    brands_mean_prices[brand] = int(mean_price)

brands_mean_prices
    
    

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

So between our top 5 brands, Audi is in the top regarding average price. VW is in between which may explain its popularity.